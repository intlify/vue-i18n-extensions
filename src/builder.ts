import { isNumber } from '@intlify/shared'

/**
 * Content Builder options
 *
 * @remarks
 * options that creating a {@link ContentBuilder}
 *
 * @public
 */
export interface ContentBuilderOptions {
  /**
   * Intdent level of Builder
   */
  indentLevel?: number
}

/**
 * Content Builder
 *
 * @public
 */
export interface ContentBuilder {
  /**
   * Indent level
   * @defaultValue 0
   */
  readonly indentLevel: number
  /**
   * Content
   */
  readonly content: string
  /**
   * Add content
   * @param content - additional content
   */
  push(content: string): void
  /**
   * Add line break
   */
  newline(): void
  /**
   * Add content with line break
   * @param content - additional content
   */
  pushline(content: string): void
  /**
   * Indent content
   * @param withNewLine - whether indent to be added without new line
   */
  indent(withoutNewLine?: boolean): void
  /**
   * DeIndent content
   * @param withoutNewLine - whether deindent to be added without new line
   */
  deindent(withoutNewLine?: boolean): void
}

/**
 * Create a Content Builder
 *
 * @param options - Content Builder options
 *
 * @returns A {@link ContentBuilder} instance
 *
 * @public
 */
export function createContentBuilder(
  options: ContentBuilderOptions = {}
): ContentBuilder {
  let _indentLevel =
    options.indentLevel != null && isNumber(options.indentLevel)
      ? options.indentLevel
      : 0
  let _content = ''

  function push(content: string): void {
    _content += content
  }

  function _newline(n: number) {
    push('\n' + `  `.repeat(n))
  }

  function newline() {
    _newline(_indentLevel)
  }

  function pushline(content: string) {
    push(content)
    newline()
  }

  function indent(withoutNewLine?: boolean) {
    if (withoutNewLine) {
      ++_indentLevel
    } else {
      _newline(++_indentLevel)
    }
  }

  function deindent(withoutNewLine?: boolean) {
    if (withoutNewLine) {
      --_indentLevel
    } else {
      _newline(--_indentLevel)
    }
  }

  return {
    get indentLevel(): number {
      return _indentLevel
    },
    get content(): string {
      return String(_content)
    },
    push,
    newline,
    pushline,
    indent,
    deindent
  }
}
