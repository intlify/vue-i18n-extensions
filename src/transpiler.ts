import { parse } from '@babel/parser'
import type { ObjectProperty, ObjectExpression, Expression } from '@babel/types'

interface EvaluateReturn {
  status: 'ok' | 'ng'
  value?: unknown
}

export type NestedValue<T> = { [P in keyof T]: NestedValue<T[P]> }

export type TranslationParams<T = Record<string, unknown>> = {
  path: string
  named: NestedValue<T>
  options: {
    locale?: string
    plural?: string
  }
}

export function evaluateValue(expression: string): EvaluateReturn {
  const ret = { status: 'ng', value: undefined } as EvaluateReturn

  try {
    const ast = parse(`const a = ${expression.trim()}`)
    if (
      ast.program.body.length >= 1 &&
      ast.program.body[0].type === 'VariableDeclaration' &&
      ast.program.body[0].declarations.length >= 1
    ) {
      const node = ast.program.body[0].declarations[0].init
      if (node != null && (node.type === 'StringLiteral' || node.type === 'ObjectExpression')) {
        // eslint-disable-next-line @typescript-eslint/no-implied-eval
        const val = new Function(`return ${expression.trim()} `)() as unknown
        ret.status = 'ok'
        ret.value = val
      }
    }
    // eslint-disable-next-line no-empty
  } catch (_e) {}

  return ret
}

export function parseVTExpression(expression: string): TranslationParams {
  const ret: TranslationParams = {
    path: '',
    named: {},
    options: {}
  }

  try {
    const ast = parse(`const a = ${expression.trim()}`)
    if (
      ast.program.body.length >= 1 &&
      ast.program.body[0].type === 'VariableDeclaration' &&
      ast.program.body[0].declarations.length >= 1
    ) {
      const node = ast.program.body[0].declarations[0].init
      if (node != null) {
        if (node.type === 'StringLiteral' && node.extra) {
          ret.path = node.extra.raw as string
        } else if (node.type === 'Identifier') {
          ret.path = node.name
        } else if (node.type === 'MemberExpression') {
          ret.path = getObjectMemberValue(node)
        } else if (node.type === 'ObjectExpression') {
          node.properties.forEach(propNode => {
            if (propNode.type === 'ObjectProperty') {
              const propKeyNode = propNode.key
              if (propKeyNode.type !== 'Identifier') {
                return
              }
              const propValueNode = propNode.value
              switch (propKeyNode.name) {
                case 'path':
                  ret.path = getObjectMemberValue(propValueNode)
                  break
                case 'locale':
                  ret.options.locale = getObjectMemberValue(propValueNode)
                  break
                case 'choice':
                case 'plural':
                  ret.options.plural = getObjectMemberValue(propValueNode)
                  break
                case 'args':
                  // console.log('args', propValueNode)
                  if (propValueNode.type === 'ObjectExpression') {
                    traverseObjectMember(propValueNode, ret.named)
                  }
                  break
                default:
                  break
              }
            }
          })
        }
      }
    }
    // eslint-disable-next-line no-empty
  } catch (_e) {}

  return ret
}

function getObjectMemberValue(node: ObjectProperty['value']): string {
  if ((node.type === 'StringLiteral' || node.type === 'NumericLiteral') && node.extra) {
    return node.extra.raw as string
  } else if (node.type === 'Identifier') {
    return node.name
  } else if (node.type === 'MemberExpression') {
    const paths: string[] = []
    collectMemberPath(node, paths)
    paths.reverse()
    return paths.join('.')
  } else {
    return ''
  }
}

function traverseObjectMember(
  node: ObjectExpression,
  target: NestedValue<Record<string, unknown>>
): void {
  node.properties.forEach(propNode => {
    if (propNode.type === 'ObjectProperty') {
      const propKeyNode = propNode.key
      if (propKeyNode.type !== 'Identifier') {
        return
      }
      if (!(propKeyNode.name in target)) {
        target[propKeyNode.name] = {}
      }
      // console.log('propNode', propNode)
      const propValueNode = propNode.value
      if (propValueNode.type === 'ObjectExpression') {
        traverseObjectMember(propValueNode, target[propKeyNode.name])
      } else {
        target[propKeyNode.name] = getObjectMemberValue(propValueNode)
      }
    }
  })
}

function collectMemberPath(node: Expression, paths: string[]): void {
  if (node.type === 'Identifier') {
    paths.push(node.name)
    return
  } else if (node.type === 'MemberExpression' && node.property.type === 'Identifier') {
    paths.push(node.property.name)
    return collectMemberPath(node.object, paths)
  }
}
