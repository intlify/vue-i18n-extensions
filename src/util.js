const stringRE = /'(?:[^'\\]|\\.)*'|"(?:[^"\\]|\\.)*"|`(?:[^`\\]|\\.)*\$\{|\}(?:[^`\\]|\\.)*`|`(?:[^`\\]|\\.)*`/g
const ecmaKeywordsRE = new RegExp('(?<!\\.)\\b' + (
  'delete,typeof,instanceof,void,do,if,for,let,new,try,var,case,else,with,await,break,catch,class,const,' +
  'alert,eval,super,throw,while,yield,delete,export,import,return,switch,default,' +
  'extends,finally,continue,debugger,function,arguments'
).split(',').join('\\b(?!\\.)|(?<!\\.)\\b') + '\\b(?!\\.)')

function warn (msg, err) {
  if (typeof console !== 'undefined') {
    console.warn('[vue-i18n-extensions] ' + msg)
    /* istanbul ignore if */
    if (err) {
      console.warn(err.stack)
    }
  }
}

const toString = Object.prototype.toString
const OBJECT_STRING = '[object Object]'
function isPlainObject (obj) {
  return toString.call(obj) === OBJECT_STRING
}

function addProp (el, name, value) {
  (el.props || (el.props = [])).push({ name, value })
}

function getAttr (el, name) {
  return el.attrsMap[name]
}

function removeAttr (el, name) {
  if ((el.attrsMap[name]) !== null) {
    const list = el.attrsList
    for (let i = 0, l = list.length; i < l; i++) {
      if (list[i].name === name) {
        list.splice(i, 1)
        break
      }
    }
  }
}

function evaluateValue (expression) {
  const ret = { status: 'ng', value: undefined }

  if (expression.match(ecmaKeywordsRE)) { return ret }
  if (!expression.match(stringRE)) { return ret }

  try {
    const val = (new Function(`return ${expression}`))()
    ret.status = 'ok'
    ret.value = val
  } catch (e) { }
  
  return ret
}

module.exports = {
  warn,
  isPlainObject,
  addProp,
  getAttr,
  removeAttr,
  evaluateValue
}
