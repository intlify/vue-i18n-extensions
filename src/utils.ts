import { parse } from '@babel/parser'

interface EvaluateReturn {
  status: 'ok' | 'ng'
  value?: unknown
}

export function evaluateValue(expression: string): EvaluateReturn {
  const ret = { status: 'ng', value: undefined } as EvaluateReturn

  try {
    const ast = parse(`const a = ${expression.trim()}`)
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const node = (ast.program.body[0] as any).declarations[0].init
    if (node.type === 'StringLiteral' || node.type === 'ObjectExpression') {
      const val = new Function(`return ${expression.trim()}`)()
      ret.status = 'ok'
      ret.value = val
    }
  } catch (e) {}

  return ret
}
