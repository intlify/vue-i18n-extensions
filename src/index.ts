import {
  DirectiveTransform,
  createObjectProperty,
  createSimpleExpression,
  SimpleExpressionNode
} from '@vue/compiler-dom'
import { isPlainObject, isString, toDisplayString } from '@vue/shared'
import { I18n, Locale } from 'vue-i18n'
import { evaluateValue } from './utils'

// TODO: should be imported from vue-i18n
type VTDirectiveValue = {
  path: string
  locale?: Locale
  args?: { [prop: string]: unknown }
  choice?: number
}

export const isNumber = (val: unknown): val is number =>
  typeof val === 'number' && isFinite(val)

export function defineTransformVT(i18n: I18n): DirectiveTransform {
  return (dir, node, context) => {
    const { exp, loc } = dir
    // console.log('v-t trans', dir)
    // console.log('v-t trans node', node)
    if (!exp) {
      // TODO: throw error
    }
    if (node.children.length) {
      // TODO: throw error
    }

    const { status, value } = evaluateValue(
      (exp as SimpleExpressionNode).content
    )
    if (status === 'ng') {
      // TODO: throw error
    }

    const parsedValue = parseValue(value)
    const translated = i18n.global.t(...makeParams(parsedValue))
    if (translated) {
      ;(exp as SimpleExpressionNode).content = toDisplayString(
        `${JSON.stringify(translated)}`
      )
    }

    return {
      props: [
        createObjectProperty(
          createSimpleExpression(`textContent`, true, loc),
          exp || createSimpleExpression('', true)
        )
      ]
    }
  }
}

// TODO: should be defined typing ...
function parseValue(value: unknown): VTDirectiveValue {
  if (isString(value)) {
    return { path: value }
  } else if (isPlainObject(value)) {
    if (!('path' in value)) {
      throw new Error('TODO')
    }
    return value as VTDirectiveValue
  } else {
    throw new Error('TODO')
  }
}

function makeParams(value: VTDirectiveValue): unknown[] {
  const { path, locale, args, choice } = value
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const options = {} as any
  const named: { [prop: string]: unknown } = args || {}

  if (isString(locale)) {
    options.locale = locale
  }

  if (isNumber(choice)) {
    options.plural = choice
  }

  return [path, named, options]
}
