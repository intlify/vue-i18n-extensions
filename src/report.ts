import { format } from './utils'
import { SourceLocation } from '@vue/compiler-core'
import { isString } from '@vue/shared'

export interface ExtensionsError extends SyntaxError {
  code: number
  loc?: SourceLocation
}

export interface ReportOptions {
  mode?: 'error' | 'warn'
  loc?: SourceLocation
  args?: unknown[]
}

export const enum ReportCodes {
  // Special value for higher-order compilers to pick up the last code
  // to avoid collision of error codes. This should always be kept as the last
  // item.
  SUCCESS,
  UNEXPECTED_DIRECTIVE_EXPRESSION,
  NOT_SUPPORTED_BINDING_PRE_TRANSLATION,
  FAILED_VALUE_EVALUATION,
  REQUIRED_PARAMETER,
  INVALID_PARAMETER_TYPE,
  NOT_SUPPORTED_PRESERVE,
  ORVERRIDE_ELEMENT_CHILDREN,
  __EXTEND_POINT__
}

// TODO: should be extracted as i18n resources (intlify project self hosting!)
const ReportMessages: { [code: number]: string } = {
  [ReportCodes.UNEXPECTED_DIRECTIVE_EXPRESSION]: `Unexpected directive expression: {0}`,
  [ReportCodes.NOT_SUPPORTED_BINDING_PRE_TRANSLATION]: `Not support binding in pre-localization: {0}`,
  [ReportCodes.FAILED_VALUE_EVALUATION]: `Failed valu evaluation: {0}`,
  [ReportCodes.REQUIRED_PARAMETER]: `Required parameter: {0}`,
  [ReportCodes.INVALID_PARAMETER_TYPE]: `Required parameter: {0}`,
  [ReportCodes.NOT_SUPPORTED_PRESERVE]: `Not supportted 'preserve': {0}`,
  [ReportCodes.ORVERRIDE_ELEMENT_CHILDREN]: `v-t will override element children: {0}`
}

export function getReportMessage(code: ReportCodes, ...args: unknown[]): string {
  return format(ReportMessages[code], ...args)
}

function createExtensionsError(
  code: ReportCodes,
  msg: string,
  loc?: SourceLocation
): ExtensionsError {
  const error = new SyntaxError(msg) as ExtensionsError
  error.code = code
  if (loc) {
    error.loc = loc
  }
  return error
}

export function report(
  code: ReportCodes,
  optinos: ReportOptions = {}
): void | ExtensionsError {
  const mode =
    optinos.mode &&
    isString(optinos.mode) &&
    ['warn', 'error'].includes(optinos.mode)
      ? optinos.mode
      : 'warn'

  const msg = getReportMessage(code, optinos.args)
  if (mode === 'warn') {
    console.warn('[vue-i18n-extensions] ' + msg)
  } else {
    // error
    throw createExtensionsError(code, msg, optinos.loc)
  }
}
