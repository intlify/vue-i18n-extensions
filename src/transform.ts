import {
  createSimpleExpression,
  NodeTypes,
  isText,
  BindingTypes,
  createCompoundExpression,
  TO_DISPLAY_STRING
} from '@vue/compiler-dom'
import {
  isNumber,
  isObject,
  isString,
  isSymbol,
  toDisplayString,
  hasOwn,
  isArray
} from '@intlify/shared'
import { evaluateValue, parseVTExpression } from './transpiler'
import { report, ReportCodes } from './report'
import { createContentBuilder } from './builder'

import type {
  DirectiveTransform,
  Node,
  SimpleExpressionNode,
  CompoundExpressionNode,
  TextNode,
  InterpolationNode,
  TransformContext
} from '@vue/compiler-dom'
import type {
  I18n,
  I18nMode,
  Locale,
  Composer,
  VueI18n,
  NamedValue,
  TranslateOptions
} from 'vue-i18n'
import type { TranslationParams, NestedValue } from './transpiler'
import type { ContentBuilder } from './builder'

// TODO: should be imported from vue-i18n
type VTDirectiveValue = {
  path: string
  locale?: Locale
  args?: { [prop: string]: unknown }
  choice?: number
  plural?: number
}

/**
 * Transform options for `v-t` custom directive
 *
 * @public
 */
export interface TransformVTDirectiveOptions<
  Messages extends Record<string, unknown> = {}, // eslint-disable-line @typescript-eslint/ban-types -- TODO: fix this
  DateTimeFormats extends Record<string, unknown> = {}, // eslint-disable-line @typescript-eslint/ban-types -- TODO: fix this
  NumberFormats extends Record<string, unknown> = {}, // eslint-disable-line @typescript-eslint/ban-types -- TODO: fix this
  Legacy extends boolean = true
> {
  /**
   * I18n instance
   *
   * @remarks
   * If this option is specified, `v-t` custom directive transform uses an I18n instance to pre-translate.
   * The translation will use the global resources registered in the I18n instance,
   * that is, `v-t` directive transform is also a limitation that the resources of each component cannot be used.
   */
  i18n?: I18n<Messages, DateTimeFormats, NumberFormats, Legacy>
  /**
   * I18n Mode
   *
   * @remarks
   * Specify the API style of vue-i18n. If you use legacy API style (e.g. `$t`) at vue-i18n, you need to specify `legacy`.
   *
   * @default 'composition'
   */
  mode?: I18nMode
  /**
   * Translation function signatures
   *
   * @remarks
   * You can specify the signature of the translation function attached to the binding context when it is codegen in the Vue Compiler.
   * If you have changed the signature to a non `t` signature in the `setup` hook or `<script setup>`, you can safely SSR it.
   * If each Vue component has a different signature for the translation function, you can specify several in an array for safe SSR.
   * This option value is `undefined` and the signature attached to the binding context is `t`.
   */
  translationSignatures?: string | string[]
}

// compatibility for this commit(v3.0.3)
// https://github.com/vuejs/vue-next/commit/90bdf59f4c84ec0af9bab402c37090d82806cfc1
const enum ConstantTypes {
  NOT_CONSTANT = 0,
  CAN_SKIP_PATCH,
  CAN_HOIST,
  CAN_STRINGIFY
}

/**
 * Transform `v-t` custom directive
 *
 * @remarks
 * Transform that  `v-t` custom directive is optimized vue-i18n code by Vue compiler.
 * This transform can improve the performance by pre-translating, and it does support SSR.
 *
 * @param options - `v-t` custom directive transform options, see {@link TransformVTDirectiveOptions}
 * @returns Directive transform
 *
 * @example
 * ```js
 * import { compile } from '@vue/compiler-dom'
 * import { createI18n } from 'vue-i18n'
 * import { transformVTDirective } from '@intlify/vue-i18n-extensions'
 *
 * // create i18n instance
 * const i18n = createI18n({
 *   locale: 'ja',
 *   messages: {
 *     en: {
 *       hello: 'hello'
 *     },
 *     ja: {
 *       hello: 'こんにちは'
 *     }
 *   }
 * })
 *
 * // get transform from  `transformVTDirective` function, with `i18n` option
 * const transformVT = transformVTDirective({ i18n })
 *
 * const { code } = compile(`<p v-t="'hello'"></p>`, {
 *   mode: 'function',
 *   hoistStatic: true,
 *   prefixIdentifiers: true,
 *   directiveTransforms: { t: transformVT } // <- you need to specify to `directiveTransforms` option!
 * })
 *
 * console.log(code)
 * // output ->
 * //   const { createVNode: _createVNode, openBlock: _openBlock, createBlock: _createBlock } = Vue
 * //   return function render(_ctx, _cache) {
 * //     return (_openBlock(), _createBlock(\\"div\\", null, \\"こんにちは！\\"))
 * //   }
 * ```
 * @public
 */
export function transformVTDirective<
  Messages extends Record<string, unknown> = {}, // eslint-disable-line @typescript-eslint/ban-types -- TODO: fix this
  DateTimeFormats extends Record<string, unknown> = {}, // eslint-disable-line @typescript-eslint/ban-types -- TODO: fix this
  NumberFormats extends Record<string, unknown> = {}, // eslint-disable-line @typescript-eslint/ban-types -- TODO: fix this
  Legacy extends boolean = true
>(
  options: TransformVTDirectiveOptions<Messages, DateTimeFormats, NumberFormats, Legacy> = {}
): DirectiveTransform {
  const i18nInstance = options.i18n
  const mode =
    isString(options.mode) && ['composition', 'legacy'].includes(options.mode)
      ? options.mode
      : 'composition'
  // prettier-ignore
  const translationSignatures = isString(options.translationSignatures)
    ? [options.translationSignatures]
    : isArray(options.translationSignatures)
      ? options.translationSignatures
      : ['t']
  translationSignatures.push(`$t`) // fallback to global scope

  return (dir, node, context) => {
    const { exp, loc } = dir
    // console.log('v-t dir', dir)
    // console.log('v-t node', node)
    if (!exp) {
      // TODO: throw error with context.OnError
      // NOTE: We need to support from @vue/compiler-core
      // https://github.com/vuejs/vue-next/issues/1147
      report(ReportCodes.UNEXPECTED_DIRECTIVE_EXPRESSION, {
        mode: 'error',
        args: [node.loc.source || ''],
        loc: node.loc
      })
    }

    if (node.children.length > 0) {
      // TODO: throw error with context.OnError
      // NOTE: We need to support from @vue/compiler-core
      // https://github.com/vuejs/vue-next/issues/1147
      report(ReportCodes.OVERRIDE_ELEMENT_CHILDREN, {
        mode: 'error',
        args: [node.loc.source || ''],
        loc: node.loc
      })
      node.children.length = 0
    }

    if (isSimpleExpressionNode(exp)) {
      if (isConstant(exp) && i18nInstance) {
        const { status, value } = evaluateValue(exp.content)
        if (status === 'ng') {
          report(ReportCodes.FAILED_VALUE_EVALUATION, {
            args: [node.loc.source || ''],
            loc: node.loc
          })
          return { props: [] }
        }

        const [parsedValue, parseStatus] = parseValue(value)
        if (parseStatus !== ReportCodes.SUCCESS) {
          report(parseStatus, { args: [node.loc.source || ''], loc: node.loc })
          return { props: [] }
        }
        if (parsedValue == null) {
          report(ReportCodes.UNEXPECTED_ERROR, {
            args: [node.loc.source || ''],
            loc: node.loc
          })
          return { props: [] }
        }

        const global = getComposer(i18nInstance as unknown as I18n)
        if (global == null) {
          report(ReportCodes.NOT_RESOLVED_COMPOSER, {
            args: [node.loc.source || ''],
            loc: node.loc
          })
          return { props: [] }
        }
        const content = global.t(...makeParams(parsedValue))

        node.children.push({
          type: NodeTypes.TEXT,
          content
        } as TextNode)
        return { props: [] }
      } else {
        const translationParams = parseVTExpression(exp.content)
        const code = generateTranslationCode(
          context,
          mode,
          translationParams,
          translationSignatures
        )
        context.helper(TO_DISPLAY_STRING)
        node.children.push({
          type: NodeTypes.INTERPOLATION,
          content: createCompoundExpression([
            createSimpleExpression(code, false, loc, ConstantTypes.NOT_CONSTANT as number)
          ])
        } as InterpolationNode)
        return { props: [] }
      }
    } else if (isCompoundExpressionNode(exp)) {
      const content = exp.children.map(mapNodeContentHandler).join('')
      const code = generateTranslationCode(
        context,
        mode,
        parseVTExpression(content),
        translationSignatures
      )
      context.helper(TO_DISPLAY_STRING)
      node.children.push({
        type: NodeTypes.INTERPOLATION,
        content: createCompoundExpression([
          createSimpleExpression(code, false, loc, ConstantTypes.NOT_CONSTANT as number)
        ])
      } as InterpolationNode)
      return { props: [] }
    } else {
      report(ReportCodes.NOT_SUPPORTED, {
        args: [node.loc.source || ''],
        loc: node.loc
      })
      return { props: [] }
    }
  }
}

function getComposer(i18n: I18n): Composer | null {
  return isI18nInstance(i18n) ? getComposerInternal(i18n) : null
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function isI18nInstance(i18n: any): i18n is I18n {
  return i18n != null && 'global' in i18n && 'mode' in i18n
}

function getComposerInternal(i18n: I18n): Composer | null {
  if (i18n.mode === 'composition') {
    return isComposer(i18n.global) ? i18n.global : null
  } else {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-member-access
    return isVueI18n(i18n.global) ? ((i18n.global as any).__composer as Composer) : null
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function isComposer(target: any): target is Composer {
  return target != null && !('__composer' in target)
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function isVueI18n(target: any): target is VueI18n {
  return target != null && '__composer' in target
}

function isSimpleExpressionNode(node: Node | undefined): node is SimpleExpressionNode {
  return node != null && node.type === NodeTypes.SIMPLE_EXPRESSION
}

function isCompoundExpressionNode(node: Node | undefined): node is CompoundExpressionNode {
  return node != null && node.type === NodeTypes.COMPOUND_EXPRESSION
}

function isConstant(node: SimpleExpressionNode): boolean {
  if ('isConstant' in node) {
    // for v3.0.3 earlier
    // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-member-access
    return (node as any).isConstant as boolean
  } else if ('constType' in node) {
    // for v3.0.3 or later
    return (node.constType as number) <= (ConstantTypes.CAN_STRINGIFY as unknown as number)
  } else {
    throw Error('unexpected error in Vue SimpleExpressionNode')
  }
}

function mapNodeContentHandler(
  value:
    | string
    | symbol
    | SimpleExpressionNode
    | CompoundExpressionNode
    | TextNode
    | InterpolationNode
): string {
  if (isString(value)) {
    return value
  } else if (isSymbol(value)) {
    return value.description || ''
  } else if (isSimpleExpressionNode(value)) {
    return value.content
  } else if (isCompoundExpressionNode(value)) {
    return value.children.map(mapNodeContentHandler).join('')
  } else if (isText(value)) {
    if (isString(value.content)) {
      return value.content
    } else if (isSimpleExpressionNode(value.content)) {
      return value.content.content
    } else if (isCompoundExpressionNode(value.content)) {
      return value.content.children.map(mapNodeContentHandler).join('')
    } else {
      return ''
    }
  } else {
    return ''
  }
}

function parseValue(value: unknown): [VTDirectiveValue | null, ReportCodes] {
  if (isString(value)) {
    return [{ path: value }, ReportCodes.SUCCESS]
  } else if (isObject(value)) {
    if (!('path' in value)) {
      return [null, ReportCodes.REQUIRED_PARAMETER]
    }
    return [value as VTDirectiveValue, ReportCodes.SUCCESS]
  } else {
    return [null, ReportCodes.INVALID_PARAMETER_TYPE]
  }
}

function makeParams(value: VTDirectiveValue): [string, NamedValue, TranslateOptions] {
  const { path, locale, args, choice, plural } = value

  const options = {} as TranslateOptions
  const named: NamedValue = args || {}

  if (isString(locale)) {
    options.locale = locale
  }

  if (isNumber(choice)) {
    options.plural = choice
  }

  if (isNumber(plural)) {
    options.plural = plural
  }

  return [path, named, options]
}

function generateTranslationCode(
  context: TransformContext,
  mode: I18nMode,
  params: TranslationParams,
  translationSignatures: string[]
): string {
  return mode === 'composition'
    ? generateComposableCode(context, params, translationSignatures)
    : generateLegacyCode(context, params)
}

function generateTranslationCallableSignatures(
  context: TransformContext,
  translationSignatures: string[]
): string {
  const { prefixIdentifiers, bindingMetadata, inline } = context
  return translationSignatures
    .map(signature => {
      const type = hasOwn(bindingMetadata, signature) && bindingMetadata[signature]
      if (inline) {
        return signature
      } else {
        const bindingContext = prefixIdentifiers
          ? (type && type.startsWith('setup')) || type === BindingTypes.LITERAL_CONST
            ? '$setup.'
            : '_ctx.'
          : ''
        return `${bindingContext}${signature}`
      }
    })
    .join(' || ')
}

function generateComposableCode(
  context: TransformContext,
  params: TranslationParams,
  translationSignatures: string[]
): string {
  const baseCode = `(${generateTranslationCallableSignatures(context, translationSignatures)})`

  const builder = createContentBuilder()
  builder.push(`${baseCode}(`)

  // generate path
  builder.push(`${toDisplayString(params.path)}`)

  // generate named
  builder.push(`, { `)
  generateNamedCode(builder, params.named)
  builder.push(` }`)

  // generate options
  builder.push(`, { `)
  if (params.options.locale) {
    builder.push(`locale: ${toDisplayString(params.options.locale)}`)
  }
  if (params.options.plural) {
    if (params.options.locale) {
      builder.push(', ')
    }
    builder.push(`plural: ${toDisplayString(params.options.plural)}`)
  }
  builder.push(` }`)

  builder.push(`)`)
  const content = builder.content
  return content
}

function generateNamedCode(
  builder: ContentBuilder,
  named: NestedValue<Record<string, unknown>>
): void {
  const keys = Object.keys(named)
  keys.forEach(k => {
    const v = named[k]
    if (isObject(v)) {
      builder.push(`${k}: { `)
      generateNamedCode(builder, v)
      builder.push(` }`)
    } else {
      builder.push(`${k}: ${toDisplayString(v)}`)
    }
  })
}

function generateLegacyCode(context: TransformContext, params: TranslationParams): string {
  const mode = !params.options.plural ? 'basic' : 'plural'
  const baseCode = `${context.prefixIdentifiers ? '_ctx.' : ''}${mode === 'basic' ? '$t' : '$tc'}`

  const builder = createContentBuilder()
  builder.push(`${baseCode}(`)

  // generate path
  builder.push(`${toDisplayString(params.path)}`)

  if (mode === 'basic') {
    // generate locale
    if (isString(params.options.locale)) {
      builder.push(`, ${toDisplayString(params.options.locale)}`)
    }
    // generate named
    builder.push(`, { `)
    generateNamedCode(builder, params.named)
    builder.push(` }`)
  } else {
    // generate plural
    builder.push(`, ${toDisplayString(params.options.plural)}`)
    if (isString(params.options.locale)) {
      // generate locale
      builder.push(`, ${toDisplayString(params.options.locale)}`)
    } else {
      // generate named
      builder.push(`, { `)
      generateNamedCode(builder, params.named)
      builder.push(` }`)
    }
  }

  builder.push(`)`)
  const content = builder.content
  return content
}
