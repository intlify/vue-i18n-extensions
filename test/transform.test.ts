import { compile } from '@vue/compiler-dom'
import { defineTransformVT } from '../src/index'
import { createI18n } from 'vue-i18n'

test('string literal', () => {
  const i18n = createI18n({
    locale: 'ja',
    messages: {
      ja: { hello: 'こんにちは！' },
      en: { hello: 'hello!' }
    }
  })
  const transformVT = defineTransformVT(i18n)
  const source = `<div v-t="'hello'"/>`
  const { code } = compile(source, {
    mode: 'function',
    hoistStatic: true,
    prefixIdentifiers: true,
    directiveTransforms: { t: transformVT }
  })
  expect(code).toMatchSnapshot(source)
})

test.todo('object literal')
test.todo('plural')
test.todo('none expression')
test.todo('invalid expression')
test.todo('preserve modifier')
test.todo('have child elements')
test.todo('missing translation')
