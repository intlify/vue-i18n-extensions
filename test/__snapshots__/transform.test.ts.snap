// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`string literal: <div v-t="'hello'"/> 1`] = `
"const { createVNode: _createVNode, openBlock: _openBlock, createBlock: _createBlock } = Vue

const _hoisted_1 = { textContent: \\"こんにちは！\\" }

return function render(_ctx, _cache) {
  return (_openBlock(), _createBlock(\\"div\\", _hoisted_1))
}"
`;
